packages/query-persist-client-core/src/createPersister.ts

```diff
@@ -0,0 +1,147 @@
+import { hashKey } from '@tanstack/query-core'
+import type {
+  QueryClient,
+  QueryFunctionContext,
+  QueryKey,
+  QueryState,
+} from '@tanstack/query-core'
+
+export interface PersistedQuery {
+  buster: string
+  timestamp: number
+  queryHash: string
+  queryKey: QueryKey
+  state: QueryState
+}
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | undefined | null>
+  setItem: (key: string, value: string) => Promise<void>
+  removeItem: (key: string) => Promise<void>
+}
+
+export interface StoragePersisterOptions<QC extends QueryClient> {
+  /**
+   * Query Client instance
+   */
+  queryClient: QC
+  /** The storage client used for setting and retrieving items from cache.
+   * For SSR pass in `undefined`.
+   */
+  storage: AsyncStorage | Storage | undefined | null
+  /**
+   * How to serialize the data to storage.
+   * @default `JSON.stringify`
+   */
+  serialize?: (client: PersistedQuery) => string
+  /**
+   * How to deserialize the data from storage.
+   * @default `JSON.parse`
+   */
+  deserialize?: (cachedString: string) => PersistedQuery
+  /**
+   * A unique string that can be used to forcefully invalidate existing caches,
+   * if they do not share the same buster string
+   */
+  buster?: string
+  /**
+   * The max-allowed age of the cache in milliseconds.
+   * If a persisted cache is found that is older than this
+   * time, it will be discarded
+   */
+  maxAge?: number
+}
+
+/**
+ * Warning: experimental feature.
+ * This utility function enables fine-grained query persistance.
+ * Simple add it as a `persister` parameter to `useQuery` or `defaultOptions` on `queryClient`.
+ * 
+ * ```
+ * useQuery({
+     queryKey: ['myKey'],
+     queryFn: fetcher,
+     persister: createPersister({
+       storage: localStorage,
+       queryClient,
+     }),
+   })
+   ```
+ */
+export function createPersister<T, QC extends QueryClient>({
+  queryClient,
+  storage,
+  buster = '',
+  maxAge = 1000 * 60 * 60 * 24,
+  serialize = JSON.stringify,
+  deserialize = JSON.parse,
+}: StoragePersisterOptions<QC>) {
+  return async (
+    queryFn: (context: QueryFunctionContext) => T | Promise<T>,
+    context: QueryFunctionContext,
+  ) => {
+    const queryHash = hashKey(context.queryKey)
+    const queryState = queryClient.getQueryState(context.queryKey)
+
+    // Try to restore only if we do not have any data in the cache and we have persister defined
+    if (!queryState?.data && storage != null) {
```

Comment:
I think we need to check for `queryState?.data !== undefined` explicitly, because having falsy values in `data` for a successful query is a legit use-case.